src/clients/google-client.ts

```diff
@@ -0,0 +1,92 @@
+import {Auth, docs_v1, drive_v3, driveactivity_v2, google, sheets_v4, slides_v1} from 'googleapis';
+import {head} from 'lodash';
+
+import {ExceptionType, KVStoreGoogleData} from '../constant';
+import {AppCallRequest, KVGoogleData, KVGoogleUser, KVStoreOptions, Oauth2App, Oauth2CurrentUser} from '../types';
+import {configureI18n} from '../utils/translations';
+import {tryPromise} from '../utils/utils';
+
+import {KVStoreClient} from '.';
+
+export const getOAuthGoogleClient = async (call: AppCallRequest): Promise<Auth.OAuth2Client> => {
+    const oauth2App: Oauth2App = call.context.oauth2 as Oauth2App;
+    const oAuth2Client = new google.auth.OAuth2(
+        oauth2App.client_id,
+        oauth2App.client_secret,
+        oauth2App.complete_url,
+    );
+
+    return oAuth2Client;
+};
+
+export const getGoogleOAuth = async (call: AppCallRequest): Promise<Auth.OAuth2Client> => {
+    const i18nObj = configureI18n(call.context);
+
+    const mattermostUrl: string | undefined = call.context.mattermost_site_url;
+    const botAccessToken: string | undefined = call.context.bot_access_token;
+    const userID: string | undefined = call.context.acting_user?.id;
+    let oauth2Token = call.context.oauth2?.user as Oauth2CurrentUser;
+
+    if (!oauth2Token?.refresh_token) {
+        const kvOptions: KVStoreOptions = {
+            mattermostUrl: <string>mattermostUrl,
+            accessToken: <string>botAccessToken,
+        };
+        const kvStoreClient = new KVStoreClient(kvOptions);
+        const googleData: KVGoogleData = await kvStoreClient.kvGet(KVStoreGoogleData.GOOGLE_DATA);
+        const kvGUser: KVGoogleUser | undefined = googleData?.userData?.find((user) => head(Object.keys(user)) === <string>userID);
+        if (Boolean(kvGUser)) {
+            oauth2Token = head(Object.values(<KVGoogleUser>kvGUser)) as Oauth2CurrentUser;
+        }
+    }
+
+    const oauth2Client = await getOAuthGoogleClient(call);
+    oauth2Client.setCredentials(oauth2Token);
+    await tryPromise(oauth2Client.refreshAccessToken(), ExceptionType.MARKDOWN, i18nObj.__('general.google-error'));
```

Comment:
Is it necessary to store the token? When we execute `oauth2Client.refreshAccessToken()` we do not have access to the token. It is stored on the Client as part of Google Library, we only get access to the refresh_token, that never changes

![token](https://user-images.githubusercontent.com/118392470/207700237-f8aace3e-73d2-4b14-883d-2d5a005bd56c.png)

Reply:
> we only get access to the refresh_token, that never changes

We've had problems with refresh tokens in other integrations, where the refresh tokens themselves were invalidated. These external integrations can change their policy at this at any time, and they do change. They often make it so you have to store a new refresh token upon refreshing the OAuth token. Are you positive that the refresh token never changes upon refreshing the access token? Once again, this could change at any time. Please take another look and see if this token is available inside of the `oauth2Client` somehow